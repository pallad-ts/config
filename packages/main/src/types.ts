import * as is from "predicates";
import {URL} from "url";
import {ConfigError} from "./ConfigError";
import yn = require('yn');

export function string(x: unknown): string {
    return (String(x)).trim();
}

export function int(x: unknown) {
    const result = parseInt(x + '', 10);
    if (is.nan(result)) {
        throw new ConfigError(`Value "${x}" cannot be converted to Int`);
    }
    return result;
}

export function number(x: unknown) {
    const result = parseFloat(x + '');
    if (is.nan(result)) {
        throw new ConfigError(`Value "${x}" cannot be converted to Number`);
    }
    return result;
}

export function bool(x: unknown): boolean {
    const result = yn(x);
    if (result === undefined) {
        throw new ConfigError(`Value "${x}" cannot be converted to bool`);
    }
    return result;
}

export function url(options: url.Options = {}) {
    if (options.protocols) {
        options.protocols = options.protocols.map(x => x.toLowerCase());
    }

    return (x: any): string => {
        let url: URL;
        try {
            url = new URL(x);
        } catch (e) {
            throw new ConfigError(e.message);
        }

        if (options.protocols) {
            const protocol = url.protocol.replace(/:$/, '');
            if (!options.protocols.includes(protocol)) {
                throw new ConfigError(`Protocol "${protocol}" is not allowed. Allowed: ${options.protocols.join(', ')}`);
            }
        }
        return x;
    }
}

export namespace url {
    export interface Options {
        protocols?: string[];
    }
}
