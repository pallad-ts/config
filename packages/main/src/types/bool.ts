import yn = require("yn");
import { ERRORS } from "../errors";

export function bool(x: unknown): boolean {
    const result = yn(x);
    if (result === undefined) {
        throw ERRORS.CANNOT_CONVERT_TO_BOOL.format(String(x));
    }
    return result;
}
