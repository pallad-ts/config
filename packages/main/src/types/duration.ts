import { fromTry, left, right } from "@sweet-monads/either";
import { parse, Duration, toSeconds } from "iso8601-duration";

import { Less, Greater, Equal, Result } from "@pallad/compare";
import { enchant, Range, Enchanted } from "@pallad/range";

import { ERRORS } from "../errors";

function durationComparator(a: Duration, b: Duration): Result {
    const aInSeconds = toSeconds(a);
    const bInSeconds = toSeconds(b);
    if (aInSeconds > bInSeconds) {
        return Greater;
    } else if (aInSeconds < bInSeconds) {
        return Less;
    }
    return Equal;
}

function factory(options: duration.Options = {}) {
    const input = [
        options.min ? toDuration(options.min) : undefined,
        options.max ? toDuration(options.max) : undefined,
    ];

    let range: Enchanted<Duration> | undefined;

    if (options.min || options.max) {
        range = enchant(Range.fromArray(input as Duration[], durationComparator), durationComparator);
    }

    return (value: unknown): Duration => {
        return fromTry<Error, Duration>(() => parse((value as string) + ""))
            .mapLeft(e => ERRORS.CANNOT_CONVERT_TO_DURATION.create(value, e.message))
            .chain(duration => {
                if (range && !range.isWithin(duration)) {
                    return left(ERRORS.DURATION_INVALID_VALUE_IN_RANGE.create(duration, range));
                }

                return right(duration);
            })
            .unwrap(e => e);
    };
}

export const duration = (() => {
    const result = factory() as duration.Factory;
    result.options = factory;
    return result;
})();

function toDuration(value: string | Duration): Duration {
    return typeof value === "string" ? parse(value) : value;
}

export namespace duration {
    export interface Factory extends Converter {
        options(options: Options): Converter;
    }

    export interface Options {
        min?: string | Duration;
        max?: string | Duration;
    }

    export interface Converter {
        (value: unknown): Duration;
    }
}
