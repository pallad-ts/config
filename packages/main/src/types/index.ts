/**
 * @file Automatically generated by barrelsby.
 */

export * from "./bool";
export * from "./int";
export * from "./number";
export * from "./split";
export * from "./string";
export * from "./url";
