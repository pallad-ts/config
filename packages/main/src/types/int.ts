import { ERRORS } from "../errors";

export function int(x: unknown) {
    const result = parseInt(String(x), 10);
    if (isNaN(result)) {
        throw ERRORS.CANNOT_CONVERT_TO_INT.format(x);
    }
    return result;
}
