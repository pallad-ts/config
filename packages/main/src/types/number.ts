import { ERRORS } from "../errors";

export function number(x: unknown) {
    const result = parseFloat(String(x));
    if (isNaN(result)) {
        throw ERRORS.CANNOT_CONVERT_TO_NUMBER.format(x);
    }
    return result;
}
