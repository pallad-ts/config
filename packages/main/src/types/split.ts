function factory<T>(options: { transformer: (value: string) => T }): (value: string) => T[];
function factory(options?: Split.Options<string>): (value: string) => string[];
function factory<T>(options?: Split.Options<T>) {
    return (value: string) => {
        let tmpResult: any[] = (value || '').split(options?.separator ?? ',')
            .map(x => x.trim())
            .filter(x => x);

        if (options?.transformer) {
            tmpResult = tmpResult.map(options.transformer).filter(x => x)
        }
        return tmpResult;
    }
}

export const split = factory({separator: ','}) as Split;

split.by = factory;

export interface Split<T = string> {
    (value: string): T[];

    by: typeof factory;
}

export namespace Split {
    export interface Options<T = string> extends Partial<Options.Transformer<T>> {
        separator?: string;
    }

    export namespace Options {
        export interface Transformer<T> {
            transformer: (value: string) => T
        }
    }
}
