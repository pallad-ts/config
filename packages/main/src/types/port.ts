import {ERRORS} from '../errors';
import {int} from './int';

function factory(options?: Port.Options) {
    const finalOptions: Required<Port.Options> = {
        registered: false,
        wellKnown: false,
        ephemeral: false,
        ...(options ?? {})
    };

    const ranges: Range[] = [];
    if (finalOptions.wellKnown) {
        ranges.push({
            description: 'well known',
            start: 1,
            end: 1023
        });
    }

    if (finalOptions.registered) {
        ranges.push({
            description: 'registered',
            start: 1024,
            end: 49151
        });
    }

    if (finalOptions.ephemeral) {
        ranges.push({
            description: 'ephemeral',
            start: 49152,
            end: 65535
        });
    }

    const description = ranges.length === 0 ? 'None, check your port type settings'
        : ranges.map(x => {
            return `${x.description} (${x.start}-${x.end})`
        }).join(', ');

    return (x: any): number => {
        const port = int(x);

        const isValid = ranges.some(x => x.start <= port && x.end >= port);

        if (!isValid) {
            throw ERRORS.INVALID_PORT_NUMBER.format(port, description);
        }

        return port;
    }
}

interface Range {
    description: string;
    start: number;
    end: number;
}

export interface Port {
    (value: any): number;

    options: typeof factory
}

export namespace Port {
    export interface Options {
        /**
         * whether to allow registered ports (1024-49151)
         */
        registered?: boolean;
        /**
         * whether to allow well known ports (1-1023)
         */
        wellKnown?: boolean;
        /**
         * whether to allow ephemeral ports (49152 - 65535)
         */
        ephemeral?: boolean
    }
}

export const port = (() => {
    const result = factory({
        wellKnown: true,
        registered: true,
    }) as any;
    result.options = factory;
    return result as Port;
})();
