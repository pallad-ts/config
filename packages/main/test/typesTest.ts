import * as validations from '@src/types';

describe('types', () => {
    describe('string', () => {
        it('gets trimmed', () => {
            expect(validations.string('test '))
                .toEqual('test');
        });

        it('converts to string', () => {
            expect(
                validations.string({
                    toString() {
                        return 'test'
                    }
                })
            )
                .toEqual('test');
        });
    });

    describe('int', () => {
        it.each([
            ['10'],
            [10],
            [10.5],
            ['10.5'],
            ['10px']
        ])('converts to int: %s', (input) => {
            expect(validations.int(input))
                .toEqual(10);
        });

        it('fails if not a number', () => {
            expect(() => {
                validations.int('foo')
            })
                .toThrowErrorMatchingSnapshot();
        });
    });

    describe('number', () => {
        it.each([
            ['10', 10],
            [10, 10],
            [10.5, 10.5],
            ['10.5', 10.5],
            ['10.5px', 10.5]
        ])('converts to number: %s', (input, expected) => {
            expect(validations.number(input))
                .toEqual(expected);
        });

        it('fails if not a number', () => {
            expect(() => {
                validations.number('foo')
            })
                .toThrowErrorMatchingSnapshot();
        });
    });

    describe('bool', () => {
        it.each<[any, boolean]>([
            ['yes', true],
            ['no', false],
            ['on', true],
            ['off', false],
            ['1', true],
            ['0', false],
            [1, true],
            [0, false]
        ])('converts to bool: %s - %s', (value, expected) => {
            expect(validations.bool(value))
                .toEqual(expected);
        });

        it('throws error if cannot be converted to bool', () => {
            expect(() => {
                validations.bool('some strange value')
            })
                .toThrowErrorMatchingSnapshot();
        });
    });

    describe('url', () => {
        describe('validation', () => {
            const validator = validations.url;

            it('success', () => {
                expect(validator('http://9marshals.com'))
                    .toEqual('http://9marshals.com')
            });

            it('failure', () => {
                expect(() => {
                    validator('test')
                })
                    .toThrowErrorMatchingSnapshot();
            })
        });

        describe('requiring protocol', () => {
            const validator = validations.url.options({protocols: ['http']});
            it('success', () => {
                expect(validator('http://9marshals.com'))
                    .toEqual('http://9marshals.com');
            });

            it('failure', () => {
                expect(() => {
                    validator('ftp://9marshals.com')
                })
                    .toThrowErrorMatchingSnapshot();
            });
        })
    });
});
