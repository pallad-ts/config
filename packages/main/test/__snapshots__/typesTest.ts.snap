// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`types bool throws error if cannot be converted to bool 1`] = `"Value "some strange value" cannot be converted to bool"`;

exports[`types int fails if not a number 1`] = `"Value "foo" cannot be converted to int"`;

exports[`types number fails if not a number 1`] = `"Value "foo" cannot be converted to number"`;

exports[`types url requiring protocol failure 1`] = `"Protocol "ftp" is not allowed. Allowed: http"`;

exports[`types url validation failure 1`] = `"Invalid URL"`;
